vmbindings/dummyvm/src/edges.rs

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+}
+
+/// Compressed OOP edge only makes sense on 64-bit architectures.
+#[cfg(target_pointer_width = "64")]
+pub mod only_64_bit {
+    use super::*;
+
+    /// This represents a location that holds a 32-bit pointer on a 64-bit machine.
+    ///
+    /// OpenJDK uses this kind of edge to store compressed OOPs on 64-bit machines.
+    #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+    pub struct CompressedOopEdge {
+        slot_addr: *mut Atomic<u32>,
+    }
+
+    unsafe impl Send for CompressedOopEdge {}
+
+    impl CompressedOopEdge {
+        pub fn from_address(address: Address) -> Self {
+            Self {
+                slot_addr: address.to_mut_ptr(),
+            }
+        }
+        pub fn as_address(&self) -> Address {
+            Address::from_mut_ptr(self.slot_addr)
+        }
+    }
+
+    impl Edge for CompressedOopEdge {
+        fn load(&self) -> ObjectReference {
+            let compressed = unsafe { (*self.slot_addr).load(atomic::Ordering::Relaxed) };
+            let expanded = (compressed as usize) << 3;
+            unsafe { Address::from_usize(expanded).to_object_reference() }
+        }
+
+        fn store(&self, object: ObjectReference) {
+            let expanded = object.to_address().as_usize();
+            let compressed = (expanded >> 3) as u32;
+            unsafe { (*self.slot_addr).store(compressed, atomic::Ordering::Relaxed) }
+        }
+    }
+}
+
+/// This represents an edge that holds a pointer to the *middle* of an object, and the offset is known.
+///
+/// Julia uses this trick to facilitate deleting array elements from the front.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct OffsetEdge {
+    slot_addr: *mut Atomic<Address>,
+    offset: usize,
+}
+
+unsafe impl Send for OffsetEdge {}
+
+impl OffsetEdge {
+    pub fn new_no_offset(address: Address) -> Self {
+        Self {
+            slot_addr: address.to_mut_ptr(),
+            offset: 0,
+        }
+    }
+
+    pub fn new_with_offset(address: Address, offset: usize) -> Self {
+        Self {
+            slot_addr: address.to_mut_ptr(),
+            offset,
+        }
+    }
+
+    pub fn slot_address(&self) -> Address {
+        Address::from_mut_ptr(self.slot_addr)
+    }
+
+    pub fn offset(&self) -> usize {
+        self.offset
+    }
+}
+
+impl Edge for OffsetEdge {
+    fn load(&self) -> ObjectReference {
+        let middle = unsafe { (*self.slot_addr).load(atomic::Ordering::Relaxed) };
+        let begin = middle - self.offset;
+        unsafe { begin.to_object_reference() }
+    }
+
+    fn store(&self, object: ObjectReference) {
+        let begin = object.to_address();
+        let middle = begin + self.offset;
+        unsafe { (*self.slot_addr).store(middle, atomic::Ordering::Relaxed) }
+    }
+}
+
+/// This edge presents the object reference itself to mmtk-core.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct ValueEdge {
```

Comment:
Will we use a type like this when a binding wants to do node enqueuing? If so, we probably should just call it `Node` or at least in the comment, state this is also known as node.

Reply:
I thought this `ValueEdge` could help node-enqueuing, but I later found it doesn't help.  I already removed all comments that mentioned "edge enqueuing" in this PR, but I want to keep this example, just to demonstrate that the developers can think out of the box when implementing `Edge`.  I'll add some comments to clarify the purpose of this example.